#1083 Overview tab / timeline

CarlosNZ opened the pull request:
Implements #1083 

This is not right yet as you can see from the screenshots (some parsing issues down the bottom -- but that's more of an issue with #1091). But I think we should sit down as a group and figure out the exact content requirements for the timeline now that we have a skeleton, rather than making a bunch of changes now that will probably get changed again.

Also -- note that the timeline would look a bit more structured in a real application as there would be a lot less entries in a single day.

<img width="939" alt="Screen Shot 2022-01-27 at 3 58 39 PM" src="https://user-images.githubusercontent.com/5456533/151284911-eebda2f0-d59a-4b29-9249-7638e53cdd4a.png">
<img width="947" alt="Screen Shot 2022-01-27 at 3 58 51 PM" src="https://user-images.githubusercontent.com/5456533/151284913-e7a324d5-100d-47f9-bd44-a6dc8f9676c2.png">

nmadruga commented:
Something that bothered me is that the Overview is always starting closed. Not sure if it should be opened instead.

Also on a chat with Jess and Adam today I said we could finish up the current work:
- Timeline fixes
- Internal comments tab
- Generated documents tab
- Permission for user to view without being the Assigner/Review

So by the end of the week we could sit down and show the new tabs, and the original idea of showing the link to each review from the Overview page and use the time to discuss next deliverables for LOQ and any improvements to this work

nmadruga commented:
Oh we also talked today about the need to initiate 2 instances one to be a Public demo server and another one for the demo presentation, both to be including this new work + being stable and adding new templates. So challenges for next week.

nmadruga commented on src/containers/Review/overview/OverviewTab.tsx:
```diff
@@ -1,27 +1,26 @@
+      {timeline && <Overview structure={fullStructure} activityLog={timeline.rawLog} />}
+      <div id="timeline">
+        <Header as="h2" textAlign="center">
+          Activity
```
Localization please

CarlosNZ replied:
Yup, and a few others. 6bc58dc2

nmadruga commented on src/containers/Review/overview/Overview.tsx:
```diff
@@ -0,0 +1,79 @@
+          <Message.Content>
+            <div className="flex-row wrap">
+              <p className="left-item">
+                <strong>Applicant: </strong>
+                {applicant}
+              </p>
+              <p className="right-item">
+                <strong>Organisation: </strong>
+                {organisation}
+              </p>
+            </div>
+            <div className="flex-row wrap">
+              <p className="left-item">
+                <strong>Started: </strong>
+                {DateTime.fromISO(started).toLocaleString()}
+              </p>
+              <p className="right-item">
+                <strong>Completed: </strong>
+                {completed && DateTime.fromISO(completed).toLocaleString()}
+              </p>
+            </div>
+            <div className="flex-row wrap">
+              <p className="left-item">
+                <strong>Stage: </strong>
+                {stage}
+              </p>
+              <p className="right-item">
+                <strong>Outcome: </strong>
+                {Outcome[outcome as ApplicationOutcome]}
+              </p>
+            </div>
+            <div className="flex-row wrap">
+              <p className="left-item">
+                <strong>Serial: </strong>
+                {serial}
+              </p>
```
All localization missing. Sorry to be picking on this, but it would be good to already add there as well.

nmadruga commented on src/containers/Review/overview/Overview.tsx:
```diff
@@ -0,0 +1,79 @@
+  console.log(activityLog)
+  const startEvent = activityLog.find((e) => e.type === 'STATUS' && e.value === 'Started')
+  const endEvent = activityLog.find((e) => e.type === 'OUTCOME' && e.value !== 'PENDING')
+  console.log('endEvent', endEvent)
```
Please remove log

nmadruga reviewed (commented):
2 overall comments:

- We need to fix the duplication of "Reviewer X assigned themselves" - I"ll see if I can help with that, also to help me understand this work better
- When all sections can we instead of list them say **All sections**?

CarlosNZ commented:
> Something that bothered me is that the Overview is always starting closed. Not sure if it should be opened instead.

I've changed it so that the current stage starts open, other stages start closed. c54f69c7

CarlosNZ commented:
Okay, I've significantly re-factored this as discussed, with all the "logging" done in the back-end, and all the "interpretation" done here. It's much more straightforward now, although there's still some complicated aspects.

Added a generic "waiting for applicant/reviewer to make changes" message that only shows when it's the current event (i.e. nothing has happened since), as we discussed.

Also added logic to ensure that the final outcome always shows up last (it might not be in the log)

Should be pretty solid now, so I'll show this version at Friday's meeting.

nmadruga commented on src/containers/Review/overview/Overview.tsx:
```diff
@@ -0,0 +1,78 @@
+                {applicant}
+              </p>
+              <p className="right-item">
+                <strong>{strings.REVIEW_OVERVIEW_OUTCOME}: </strong>
```
```suggestion
                <strong>{strings. REVIEW_OVERVIEW_ORG}: </strong>
```

nmadruga reviewed (commented):
I agree now with moving the logic, seems more straightforward

nmadruga commented:
problem found while working of issue #1087 

Log aren't showing correctly after un-assignment (or re-assignment that would also run unassign first and then assign).
<img width="814" alt="Screen Shot 2022-02-03 at 3 22 47 PM" src="https://user-images.githubusercontent.com/16461988/152271022-dcb967df-b45d-40e6-9a79-1d69e6bf3905.png">

Snapshot: 
[unassigned_missing_log.zip](https://github.com/openmsupply/application-manager-web-app/files/7991360/unassigned_missing_log.zip)

CarlosNZ commented:
> Log aren't showing correctly after un-assignment (or re-assignment that would also run unassign first and then assign).

Okay, so as I mentioned IRL, your log is out of date so the value "STATUS = Started" is no longer used (or interpreted):
![Screen Shot 2022-02-03 at 4 32 47 PM](https://user-images.githubusercontent.com/5456533/152277018-bae6139a-5bfd-4868-bb2a-c3b6a80e5f1c.png)

However, the "Unassigned" one was a typo on my part (I had "UnAssigned" in front-end), so I've fixed that. 14020929

Good spotting, thanks.
